File: gcc/testsuite/gcc.dg/loop-versioning-10.c
/* { dg-options "-O3 -fdump-tree-lversion-details" } */

/* Test that we can version a gather-like operation in which a variable
   stride is applied to the index.  */

int
f1 (int *x, int *index, int step, int n)
{
  int res = 0;
  for (int i = 0; i < n; ++i)
    res += x[index[i] * step];
  return res;
}

int
f2 (int *x, int *index, int step, int n)
{
  int res = 0;
  for (int i = 0; i < n; ++i)
    {
      int *ptr = x + index[i] * step;
      res += *ptr;
    }
  return res;
}

int x[1000];

int
g1 (int *index, int step, int n)
{
  int res = 0;
  for (int i = 0; i < n; ++i)
    res += x[index[i] * step];
  return res;
}

int
g2 (int *index, int step, int n)
{
  int res = 0;
  for (int i = 0; i < n; ++i)
    {
      int *ptr = x + index[i] * step;
      res += *ptr;
    }
  return res;
}

/* { dg-final { scan-tree-dump-times {address term [^\n]* \* loop-invariant} 4 "lversion" } } */
/* { dg-final { scan-tree-dump-times {want to version containing loop} 4 "lversion" } } */
/* { dg-final { scan-tree-dump-times {versioned this loop} 4 "lversion" } } */

File: gcc/testsuite/gfortran.dg/loop_versioning_10.f90
! { dg-options "-O3 -fdump-tree-lversion-details" }

subroutine f1(x)
  real :: x(:, :)
  x(:, 1) = 100
end subroutine f1

subroutine f2(x, i)
  real :: x(:, :)
  integer :: i
  x(:, i) = 100
end subroutine f2

subroutine f3(x)
  real :: x(:, :)
  do j = lbound(x, 1), ubound(x, 1)
     x(j, 1) = 100
  end do
end subroutine f3

subroutine f4(x, i)
  real :: x(:, :)
  integer :: i
  do j = lbound(x, 1), ubound(x, 1)
     x(j, i) = 100
  end do
end subroutine f4

! { dg-final { scan-tree-dump-times {likely to be the innermost dimension} 4 "lversion" } }
! { dg-final { scan-tree-dump-times {want to version} 4 "lversion" } }
! { dg-final { scan-tree-dump-times {versioned} 4 "lversion" } }
